### ISSUE TITLE
Data Race in parse_dot_data.

### ISSUE THREAD

**Issue by wildoranges** [pydot/pydot#401]
in function `parse_dot_data`,  a global variable `top_graphs` is used. This global variable will be overwritten repeatedly in multithreading situations.  It is a data race and may produce unexpected results.

https://github.com/pydot/pydot/blob/1270c6d8e7e77e8dbdad460d22d6448658893717/src/pydot/dot_parser.py#L499-L516

The following is an example used with `ThreadPoolExecutor` that may cause data race if the user does not know the implementation of pydot(the call chain is `graph_from_dot_file` -> `graph_from_dot_data` -> `parse_dot_data`):
```python
import pydot
from tqdm import tqdm
from concurrent.futures import ThreadPoolExecutor, as_completed

with ThreadPoolExecutor(max_workers=max_workers) as executor:
    futures = [executor.submit(pydot.graph_from_dot_file,  dot_file) for dot_file in file_list]
    for future in tqdm(as_completed(futures), total=len(futures))
        future.result()
```

**Comment by lkk7** [comment]
Thanks for reporting.

This is a legacy thing remaining from the way `pydot` was written long ago.

I think the simplest option is to turn `top_graphs` into a thread-local variable.

**Comment by ferdnyc** [comment]
Yeah, the entire parser is implemented in the fashion of graphviz's own, which is notoriously not thread-safe and practically _made_ of global variables.

Personally, I'd prefer to see the parser in `dot_parser.py` turned into a class, with each parsing run performed by a new instance of that class, all of the functions turned into class methods, and all of the globals migrated to instance variables. (`graphparser` itself, the actual pyparsing parser definition, is also a global, and since it maintains state I worry about its reentrancy as well.)

**Comment by ferdnyc** [comment]
(It'd also make debugging f---loads easier if we could actually access all of the components of the parser, from outside the module. Right now, everything except the global `graphparser` is a local variable inside the `graph_definition()` function, making it hard to get at, say, the `node_stmt` or `attr_list` definition in isolation. The first thing I usually do when working in the parser code is rip out the entire `graph_parser` function definition and pull its contents up to the module level where I can see it all.)

**Comment by lkk7** [comment]
> `dot_parser.py` turned into a class

That was my first thought. But I didn't want to break APIs again.

But now I realize that `dot_parser.py` is more internal than I remembered.
I'm sure some unlucky legacy user will have their code broken because they use an obscure function from `dot_parser` directly, but that's something we can accept.

**Comment by ferdnyc** [comment]
> Personally, I'd prefer to see the parser in `dot_parser.py` turned into a class, with each parsing run performed by a new instance of that class, all of the functions turned into class methods, and all of the globals migrated to instance variables. (`graphparser` itself, the actual pyparsing parser definition, is also a global, and since it maintains state I worry about its reentrancy as well.)

So, my first attempt at doing that not _only_ broke, but it broke the OLD parser as well. (I copied `dot_parser.py` to a new file, class-ified the parser, and then made it conditional in `pydot/__init__.py` which parser would be used. I was very surprised that that alone caused it to break.)

That's OK, tho, turns out _that_ breakage was just the code to test  the logging, interacting badly with unittest-parallel because it reloads all of the classes. If I _either_ ripped out the logging test, OR switched from unittest-parallel to regular unittest, I was back to things working OK with the OLD `dot_parser.py` code.

However, my new code still breaks with random `ERROR` results from various tests.

Turns out pyparsing is simply not thread-safe, itself (pyparsing/pyparsing#89, and if we want to make it possible to call the parser from multiple threads we'll have to put a lock on it so that only one thread at a time can be in the parser. There's no such lock in pyparsing because, as pyparsing's lead dev @ptmcg says in that issue,

> > if we protect the whole of pyparsing with a lock
> 
> This is something that is more easily/cleanly done in the calling code than in pyparsing itself, especially since there are so few parse-time entry points.

The same may be true here, as well. We can possibly make `dot_parser` safe for multiple threads, but we'd be doing it by locking the code so that only one thread at a time can be in it. It may be easier to just _not_ try calling from multiple threads at a time, because that's not going to work, and it seems there's nothing pydot can do to make that work regardless.

**Comment by ferdnyc** [comment]
(The other possibility is, we could simply `copy.deepcopy` and reset the global `top_graphs` before returning it from each `parse_dot_data()` call, so that the results of previous parsing runs won't be impacted by any future runs. It'll slow things down slightly, but may be enough thread-safety for most users. Which is good because it's also apparently as much thread-safety as we can provide.)

**Comment by ferdnyc** [comment]
We might not even need a deepcopy, a shallow copy might be sufficient come to think of it.

**Comment by ferdnyc** [comment]
Aaand, **#TIL** that if you `copy.deepcopy` a `pydot.Dot` object, it loses all of its instance variables!

Not sure _why_, yet, but clearly a bug. Possibly in our pickling code.

But it definitely breaks things, when `pydot.Dot.create()` can't find `self.prog` because it _DOESN'T EXIST_.

**Comment by ferdnyc** [comment]
Turns out `top_graphs` doesn't need to be global at all. #404 eliminates it, with no ill effects. (However, this WON'T overcome pyparsing's lack of thread safety and magically make `dot_parser` thread safe.)

**Comment by ferdnyc** [comment]
(The `deepcopy` digression was a dead end — you can't deepcopy `frozendict`s, by design. So that went nowhere. But it wasn't necessary anyway, and I'm glad I tried it because it exposed the broken copying/pickling of `pydot.Dot` objects.)

**Comment by lkk7** [comment]
Well, if pyparsing did not set a lock beacuse 
> This is something that is more easily/cleanly done in the calling code than in pyparsing itself, especially since there are so few parse-time entry points.

Then we can take the same position. `parse_dot_data` is the only dangerous place and we could just strongly annotate the fact that the function is not thread-safe.
